Improve: Migrate C++ regex engine from PCRE to PCRE2

vadi2 opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Replace the deprecated PCRE library with its successor PCRE2 for all C++ regex operations in triggers and aliases. This modernizes the regex engine while maintaining full API compatibility with existing Lua scripts.
#### Motivation for adding to Mudlet
PCRE2 provides better Unicode support, improved performance, and is actively maintained, while PCRE (version 1) is end-of-life.
#### Other info (issues closed, discussion etc)

claude committed:
Infrastructure: Migrate C++ regex engine from PCRE to PCRE2

Replace the deprecated PCRE library with its successor PCRE2 for all C++
regex operations in triggers and aliases. This modernizes the regex engine
while maintaining full API compatibility with existing Lua scripts.

Key changes:
- Add cmake/FindPCRE2.cmake module for library discovery
- Update CMakeLists.txt to link against PCRE2 instead of PCRE
- Convert pcre_compile() calls to pcre2_compile()
- Convert pcre_exec() calls to pcre2_match() with match_data
- Update pcre_fullinfo() to pcre2_pattern_info() for named groups
- Use PCRE2-specific flags (PCRE2_UTF, PCRE2_UCP, etc.)
- Implement proper memory management with pcre2_match_data lifecycle

PCRE2 provides better Unicode support, improved performance, and is actively
maintained, while PCRE (version 1) is end-of-life.

vadi2 commented:
It compiles, but needs testing to see if it works. Could do with unit tests too.

SlySven commented:
I think that `MAX_CAPTURE_GROUPS` is now cruft - but is still used in a pair of error messages they begin with `"[Trigger Error:] ..."` in `./src/TTrigger.cpp` and `"[Alias Error:] ..."` in `./src/TAlias.cpp` - and `#define`d in `./src/TALias.h`.

ZookaOnGit commented:
add some regex tests in #8537, they all pass in current dev
one fails with this PR, capturing an optional capture group

current dev;
`matches = { "Name: Tester", "Tester" }`

this PR;
`matches = { "Name: Tester", "Tester", "" }`

vadi2 committed:
Merge branch 'development' into claude/update-pcre2-regex-013b6S5tgXQ7EJPySfqAqNRm

claude committed:
Remove obsolete MAX_CAPTURE_GROUPS and ovector overflow checks

With PCRE2, match data is dynamically allocated based on the pattern via
pcre2_match_data_create_from_pattern(), so the fixed MAX_CAPTURE_GROUPS
limit and rc==0 overflow checks are no longer applicable.

PCRE2's pcre2_match() returns:
- Positive: number of captured groups (including overall match)
- Negative: error code (e.g., PCRE2_ERROR_NOMATCH)

The old PCRE1 behavior where rc==0 indicated ovector overflow doesn't
occur with PCRE2's dynamic allocation model.

vadi2 commented:
Looks like the tests passed here but one should have failed, right?

ZookaOnGit commented:
Yes, one should have failed.  Here's my output from a manual run.

```
12 successes / 1 failure / 0 errors / 0 pending : 0.030281 seconds

Failure -> ...Workspace/Mudlet/src/mudlet-lua/tests/Regex_spec.lua @ 93
PCRE regex cases with tempRegexTrigger wont capture an optional group when not present
...Workspace/Mudlet/src/mudlet-lua/tests/Regex_spec.lua:109: Expected objects to be the same.
Passed in:
(string) ''
Expected:
type nil
```

But they all pass on latest dev.

vadi2 committed:
Update CI workflows to install PCRE2 instead of PCRE

vadi2 committed:
Restore PCRE2_FOUND setting in FindPCRE2.cmake

vadi2 commented:
Hm, in my local build of the PR I get:

13 successes / 0 failures / 0 errors / 0 pending : 0.015942 seconds

I'll check the  AppImage when that comes up.

vadi2 commented on cmake/FindPCRE2.cmake:
```diff
@@ -0,0 +1,115 @@
+mark_as_advanced(PCRE2_INCLUDE_DIR PCRE2_LIBRARY PCRE2_LIBRARY_RELEASE
+                 PCRE2_LIBRARY_DEBUG)
+
+get_filename_component(PCRE2_FILENAME ${PCRE2_LIBRARY} NAME)
+string(FIND ${PCRE2_FILENAME} .a PCRE2_STATIC)
```
Don't think we need this right now.

ZookaOnGit commented:
Well that's a good thing I guess!  What could be happening though, different libraries?

```
zooka@surface:~$ dpkg -l |grep pcre2
ii  libpcre2-16-0:amd64                                 10.46-1~deb13u1                               amd64        New Perl Compatible Regular Expression Library - 16 bit runtime files
ii  libpcre2-32-0:amd64                                 10.46-1~deb13u1                               amd64        New Perl Compatible Regular Expression Library - 32 bit runtime files
ii  libpcre2-8-0:amd64                                  10.46-1~deb13u1                               amd64        New Perl Compatible Regular Expression Library- 8 bit runtime files
ii  libpcre2-dev:amd64                                  10.46-1~deb13u1                               amd64        New Perl Compatible Regular Expression Library - development files
ii  libpcre2-posix3:amd64                               10.46-1~deb13u1                               amd64        New Perl Compatible Regular Expression Library - posix-compatible runtime files
ii  lua-rex-pcre2:amd64                                 2.9.2-3+b1                                    amd64        Perl regular expressions library (PCRE2) for the Lua language
zooka@surface:~$ 
```

vadi2 commented:
I have this:

```
$ dpkg -l |grep pcre2
ii  libpcre2-16-0:amd64                            10.42-4ubuntu2.1                         amd64        New Perl Compatible Regular Expression Library - 16 bit runtime files
ii  libpcre2-32-0:amd64                            10.42-4ubuntu2.1                         amd64        New Perl Compatible Regular Expression Library - 32 bit runtime files
ii  libpcre2-8-0:amd64                             10.42-4ubuntu2.1                         amd64        New Perl Compatible Regular Expression Library- 8 bit runtime files
ii  libpcre2-8-0:i386                              10.42-4ubuntu2.1                         i386         New Perl Compatible Regular Expression Library- 8 bit runtime files
ii  libpcre2-dev:amd64                             10.42-4ubuntu2.1                         amd64        New Perl Compatible Regular Expression Library - development files
ii  libpcre2-posix3:amd64                          10.42-4ubuntu2.1                         amd64        New Perl Compatible Regular Expression Library - posix-compatible runtime files
```

The builds are currently failing but that's because I need to change it to load pcre2 in Lua instead of pcre.

vadi2 committed:
Update rex_pcre to rex_pcre2 for PCRE2 library

vadi2 commented:
Ahh, CI and me were running with pcre most likely a s not pcre2.

ZookaOnGit commented:
Several LLMs point out "PCRE2 returns all defined capture groups regardless of whether they participated in the match. Unmatched groups will contain empty strings."  So the behaviour I am seeing is actually correct in PCRE2; it keeps the order of the captures (which actually makes sense).

I am trying to think of a scenario where this would affect end users, as they would have had to check for the existence of the matches[] index before trying to access the data anyways.

So we either remove empty captures post regex processing, or run with the new behaviour.

vadi2 commented:
What do you reckon about this?

ZookaOnGit commented:
possibly breaking change; I'd like to check some packages in the repo to see how users are handling these optional cases

lets not introduce this into 4.20 though, we'll need thorough field testing firstly

SlySven commented on src/TAlias.cpp:
```diff
@@ -157,52 +156,44 @@ bool TAlias::match(const QString& haystack)
             }
             ovector[1] = start_offset + 1;
             continue;
         } else if (rc < 0) {
```
:thinking: You know about the "else-after-return" anti-pattern - won't there be a matching "else-after-continue" (and before this PR) "else-after-goto" ones?

vadi2 replied:
https://github.com/Mudlet/Mudlet/pull/8575

SlySven reviewed (commented):
Oh, *sugar*, this change is gonna screw with our *backwards-compatibility* promise isn't it? 

Scripts/packages using the `lrexlib-pcre` module (loaded in `TLuaInterpreter.cpp`'s `TLuaInterpreter::initLuaGlobals(...)` by `require`-ing `rex_pcre`  are not going to work and we'll get run-time errors of the form `[ ERROR ] -` ... `Some functions may not be available.` AND the functions available from PCRE2 are **NOT** identical - (we'll need to change lines in the CI/CB scripts that say things like):
```sh
luarocks --lua-version 5.1 install lrexlib-pcre
```
to
```sh
luarocks --lua-version 5.1 install lrexlib-pcre2
```

It would be possible to revise the code in that function `TLuaInterpreter::initLuaGlobals(...)` (c.f. the handling of the two different `zip` modules we support) to try and `require` `rex-pcre2` before (or after) `rex-pcre` and to assign the module name `rex-pcre`  to whichever one gets loaded (which might work for SOME functions in those libraries that have the same name/arguments/returns) but since they are not identical it can still cause borkage for the end-user.

ZookaOnGit commented:
> What do you reckon about this?

After further testing and consultation; any non-matching optional captures added to the start or middle of a regex follow earlier pcre behaviour.  It is only at the ones at the end (as described above) which differ.

e.g., in matching words and optionally numbers that don't match
``` 
matches = { "abc 123 def 456", "abc", "", "def" }       -- current pcre3 behaviour
matches = { "abc 123 def 456", "abc", "", "def", "" }   -- newer pcre2 behaviour
```

I move we run with the default pcre2 behaviour in PTBs to see if there is any widespread problems.  If problems arise we simply strip 0 length values from the end of matches.

vadi2 committed:
Use update-pcre2-bundling installers branch for testing

vadi2 committed:
Merge branch 'development' into claude/update-pcre2-regex-013b6S5tgXQ7EJPySfqAqNRm

SlySven commented:
So what seems to be missing now is just a Lua wrapper to fake the `fullinfo()` function from the `lrex-pcre` library from the `lrex-pcre2` one's `patterninfo()` so existing scripts that use the former won't be broken by this change.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -5640,7 +5640,7 @@ void TLuaInterpreter::initLuaGlobals()
     }
 
-    loadLuaModule(modLoadMessageQueue, QLatin1String("rex_pcre"), tr("Some functions may not be available."));
+    loadLuaModule(modLoadMessageQueue, QLatin1String("rex_pcre2"), tr("Some functions may not be available."));
```
I was thinking more along the lines of trying both rocks - in case the user still has the old one around (like I have) - in either case the library would appear in the `rexlib_pcre` namespace  (so existing references to that in old scripts/packages will continue to work) - though we'd still need a wrapper to provide `fullinfo()` from `patterninfo()` if the former function is not seen in the library that ***is*** loaded...
```suggestion
    loaded = loadLuaModule(modLoadMessageQueue, QLatin1String("rex_pcre"));
    if (!loaded) {
         loadLuaModule(modLoadMessageQueue, QLatin1String("rex_pcre2"), tr("Some regular expression functions may not be available."), QString(), qsl("rex_pcre"));
    }

```

vadi2 committed:
Update src/TLuaInterpreter.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
The Lua windows tests are also breaking, but it is hard to see what is the problem in their output. Need to diff a successful run with the one here.

claude committed:
Fix Windows and macOS builds to use PCRE2 packages

Update packaging scripts to install and use PCRE2 instead of PCRE1:
- Windows: Install mingw-w64-pcre2 instead of mingw-w64-pcre
- Windows: Install lrexlib-pcre2 instead of lrexlib-pcre (Lua bindings)
- macOS: Install pcre2 via Homebrew instead of pcre

This aligns the build dependencies with the C++ code migration to PCRE2.
The rex_pcre2.dll packaging was already updated in commit 17d380f.

vadi2 commented:
Ready for review.

vadi2 committed:
Merge branch 'development' into claude/update-pcre2-regex-013b6S5tgXQ7EJPySfqAqNRm

ZookaOnGit commented:
no problems compiling, but getting Lua errors on profile load
same with AppImage

SlySven commented:
A small detail - but the 3rd party licence details will need to be revised to talk about the PCRE2 library instead of the PCRE one (both MIT I believe they were).

claude committed:
Update third-party license text to reference PCRE2

Updated the lrexlib license attribution in the About dialog to reference
PCRE2 instead of PCRE, reflecting the migration to the PCRE2 library.

vadi2 committed:
Merge branch 'development' into claude/update-pcre2-regex-013b6S5tgXQ7EJPySfqAqNRm

vadi2 commented:
This will not be in 4.20, will be in 4.21.
